# Abundant genetic variation is retained in laboratory schistosome populations

__Authors:__ Kathrin S. JUTZELER1,2*, Roy N. PLATT3, Robbie DIAZ4, Madison MORALES4, Winka LE CLEC’H1, Frédéric D. CHEVALIER1, Timothy J.C. ANDERSON3*

__Affiliation:__
1 Host parasite Interaction program, Texas Biomedical Research Institute, P.O. Box 760549, 78245 San Antonio, Texas, USA.
2 UT Health, Microbiology, Immunology & Molecular Genetics, San Antonio, TX 78229   
3 Disease Intervention and Prevention program, Texas Biomedical Research Institute, P.O. Box 760549, 78245 San Antonio, Texas, USA.
4 Texas Biomedical Research Institute, P.O. Box 760549, 78245 San Antonio, Texas, USA.


## ABSTRACT
Schistosomes are obligate sexual blood flukes with a complex life cycle that can be maintained in the laboratory using freshwater snails as intermediate vector and rodents as definitive hosts. However, the genetic composition of laboratory schistosome populations is poorly understood: whether genetic variation has been lost due to serial inbreeding or retained is unclear. We sequenced the genomes of 19 – 24 parasites from each of five laboratory _Schistosoma mansoni_ populations and compared these with published exome data from four S. mansoni field populations. We found abundant genomic variation (1.20-1.54 million variants) within each laboratory population. Laboratory populations retained on average 68% (π = 6.16e-04 – 8.30e-04) of the nucleotide diversity observed in the four field parasite populations (π = 8.19e-04 – 1.66e-03). However, the pattern of variation was very different in laboratory and field parasite populations. Tajima’s D was positive in all laboratory populations, indicative of recent population bottlenecks, but negative in all field populations. Current effective population size estimates of laboratory populations were lower (15 – 259), compared to field populations (3,120 – infinity); the distance between markers at which linkage disequilibrium (LD) decayed to 0.5 was longer in laboratory populations (74 – 263 kb) compared to field populations (170 bp – 27 kb). The abundant genetic variation present in laboratory schistosome populations has several important implications: (i) measurement of parasite phenotypes, such as drug resistance or immunogenicity, using laboratory parasite populations will determine average values and underestimate trait variation; (ii) genome wide association studies (GWAS) can be conducted in laboratory schistosome populations by measuring phenotypes and genotypes of individual worms; (iii) genetic drift may lead to divergence in schistosome populations maintained in different laboratories. We conclude that the abundant genetic variation retained within laboratory schistosome populations provides valuable, untapped opportunities for schistosome research.

__KEY WORDS:__ Schistosoma mansoni, genomic variation, genetic diversity, effective population size, linkage disequilibrium, genome wide association studies (GWAS)
